File: tests/testthat/test_gradient.R
context("Tests for initial=\"gradient\"")

# initial="gradient" solves for the initial state by least squares. It fixes a
# backcasting divergence for seasonal additive ETS with a trend and a large
# seasonal smoothing parameter, and must leave the cases backcasting already
# handles well unchanged, and fall back safely for unsupported specifications.

# Deterministic additive seasonal series with trend (worst-cell persistence)
m <- 12
s0 <- 10 * sin(2 * pi * (1:m) / m); s0 <- s0 - mean(s0)
set.seed(9282)
xSeas <- sim.es("AAA", obs = 120, frequency = m,
                persistence = c(0.6, 0.15, 0.35),
                initial = c(100, 2), initialSeason = s0)
sseTrue <- sum(xSeas$residuals^2)

test_that("gradient fixes the seasonal-trend backcasting divergence", {
    fB <- adam(xSeas$data, "AAA", persistence = c(0.6, 0.15, 0.35),
               initial = "backcasting", silent = TRUE)
    fG <- adam(xSeas$data, "AAA", persistence = c(0.6, 0.15, 0.35),
               initial = "gradient", silent = TRUE)
    deficitB <- (sum(residuals(fB)^2) - sseTrue) / 100
    deficitG <- (sum(residuals(fG)^2) - sseTrue) / 100
    # backcasting diverges badly here; gradient must land near -rank (13), i.e.
    # a small negative deficit. Assert it is healthy and far below backcasting.
    expect_gt(deficitB, 100)
    expect_lt(deficitG, 5)
    expect_equal(fG$initialType, "gradient")
})

test_that("gradient does not disturb a well-behaved seasonal case", {
    # Low-persistence DGP: backcasting is already healthy here (deficit ~ -rank).
    persLow <- c(0.1, 0.05, 0.1)
    set.seed(9282)
    xLow <- sim.es("AAA", obs = 120, frequency = m, persistence = persLow,
                   initial = c(100, 2), initialSeason = s0)
    sseLow <- sum(xLow$residuals^2)
    fB <- adam(xLow$data, "AAA", persistence = persLow, initial = "backcasting", silent = TRUE)
    fG <- adam(xLow$data, "AAA", persistence = persLow, initial = "gradient", silent = TRUE)
    dB <- (sum(residuals(fB)^2) - sseLow) / 100
    dG <- (sum(residuals(fG)^2) - sseLow) / 100
    # both healthy; gradient must be no worse than backcasting, and negative
    expect_lt(dG, dB + 2)
    expect_lt(dG, 0)
})

test_that("gradient works for non-seasonal additive ETS", {
    set.seed(41)
    y <- ts(cumsum(rnorm(80, 0.2, 1)) + 100)
    fG <- adam(y, "AAN", initial = "gradient", silent = TRUE)
    expect_equal(fG$initialType, "gradient")
    expect_true(is.finite(logLik(fG)))
})

test_that("gradient handles multiplicative / mixed ETS via Gauss-Newton", {
    # Positive multiplicative-seasonal series (low persistence keeps it positive).
    set.seed(42)
    s0m <- 1 + 0.15 * sin(2 * pi * (1:12) / 12)
    xM <- sim.es("MAM", obs = 120, frequency = 12, persistence = c(0.2, 0.05, 0.1),
                 initial = c(100, 1.002), initialSeason = s0m)
    skip_if(any(!is.finite(xM$data)) || min(xM$data) <= 0)
    sseM <- sum(xM$residuals^2)
    fB <- adam(xM$data, "MAM", persistence = c(0.2, 0.05, 0.1),
               initial = "backcasting", silent = TRUE)
    fG <- adam(xM$data, "MAM", persistence = c(0.2, 0.05, 0.1),
               initial = "gradient", silent = TRUE)
    # Multiplicative models have no backcasting divergence; gradient must run
    # (initType set) and be no worse than backcasting.
    expect_equal(fG$initialType, "gradient")
    expect_true(is.finite(logLik(fG)))
    dB <- (sum(residuals(fB)^2) - sseM) / 100
    dG <- (sum(residuals(fG)^2) - sseM) / 100
    expect_lt(dG, dB + 1)
})

test_that("gradient falls back to backcasting for unsupported specifications", {
    # ARIMA is out of scope (no ARIMA/xreg/multi-seasonality yet): the fit must
    # fall back to backcasting (identical result), even though the requested
    # initial label stays "gradient".
    set.seed(1)
    y <- ts(cumsum(rnorm(80, 0.1, 1)) + 100)
    fB <- adam(y, "NNN", orders = list(ar = c(1), i = c(1), ma = c(1)),
               initial = "backcasting", silent = TRUE)
    fG <- suppressWarnings(adam(y, "NNN", orders = list(ar = c(1), i = c(1), ma = c(1)),
                                initial = "gradient", silent = TRUE))
    expect_equal(logLik(fG), logLik(fB), tolerance = 1e-8)
})

test_that("gradient estimates persistence jointly (nested, no backcasting)", {
    # Persistence not provided: the optimiser estimates it while the initials
    # are solved by gradient each evaluation. Must recover a healthy fit on the
    # otherwise-divergent seasonal-trend case.
    fG <- adam(xSeas$data, "AAA", initial = "gradient", silent = TRUE)
    expect_equal(fG$initialType, "gradient")
    expect_true(is.finite(logLik(fG)))
    expect_lt((sum(residuals(fG)^2) - sseTrue) / 100, 5)
})

# The loss-aware solve: initial="gradient" profiles the actual estimation loss
# (not the SSE surrogate), so at fixed persistence the gradient fit must never
# lose to backcasting on the loss itself.

set.seed(41)
xLoss <- ts(100 + cumsum(rnorm(72, 0, 2)) +
                rep(c(5, -3, 2, -4, 6, -6, 3, -1, 2, -2, 4, -6), 6), frequency = 12)
xLossM <- ts((100 + cumsum(rnorm(72, 0, 2))) *
                 rep(c(1.05, 0.97, 1.02, 0.96, 1.06, 0.94,
                       1.03, 0.99, 1.02, 0.98, 1.04, 0.94), 6), frequency = 12)

test_that("gradient profiles the robust one-step losses (MAE, HAM)", {
    for (loss in c("MAE", "HAM")) {
        fB <- adam(xLoss, "AAA", initial = "backcasting", loss = loss, silent = TRUE)
        fG <- adam(xLoss, "AAA", initial = "gradient", loss = loss, silent = TRUE)
        expect_lt(fG$lossValue, fB$lossValue)
    }
})

test_that("gradient handles exact zero residuals in the MAE weights", {
    # A locally constant series produces exact-zero residuals: the IRLS sweep
    # must drop those rows (no epsilon floor) and stay finite.
    yTies <- ts(rep(c(100, 100, 100, 105), 20), frequency = 4)
    fG <- adam(yTies, "ANA", initial = "gradient", loss = "MAE", silent = TRUE)
    expect_true(is.finite(fG$lossValue))
})

test_that("gradient profiles the multiplicative likelihoods", {
    # dgamma (the multiplicative default), dlnorm and dinvgauss get their exact
    # rho in the Gauss-Newton branch; the likelihood must not lose to backcasting.
    for (dist in c("default", "dlnorm", "dinvgauss")) {
        args <- list(data = xLossM, model = "MAM", initial = "backcasting", silent = TRUE)
        if (dist != "default") { args$distribution <- dist }
        fB <- do.call(adam, args)
        args$initial <- "gradient"
        fG <- do.call(adam, args)
        expect_lt(fG$lossValue, fB$lossValue + 1e-8)
    }
})

test_that("gradient solves the additive multistep losses on the multistep design", {
    for (loss in c("MSEh", "TMSE", "GTMSE", "MSCE", "GPL")) {
        fB <- adam(xLoss, "AAA", initial = "backcasting", loss = loss, h = 6,
                   holdout = FALSE, silent = TRUE)
        fG <- adam(xLoss, "AAA", initial = "gradient", loss = loss, h = 6,
                   holdout = FALSE, silent = TRUE)
        expect_lt(fG$lossValue, fB$lossValue)
    }
})

test_that("gradient with a custom loss function falls back to backcasting", {
    lossCustom <- function(actual, fitted, B) { sum(abs(actual - fitted)^1.5) }
    fB <- adam(xLoss, "AAA", initial = "backcasting", loss = lossCustom, silent = TRUE)
    fG <- suppressMessages(adam(xLoss, "AAA", initial = "gradient",
                                loss = lossCustom, silent = TRUE))
    expect_equal(fG$lossValue, fB$lossValue, tolerance = 1e-8)
})

File: python/tests/test_gradient.py
"""Tests for ``initial="gradient"`` — the least-squares initial-state solve.

Gradient initialisation solves for the initial ETS state that minimises the
in-sample SSE at the current persistence (exact one-shot least squares for
additive models, Gauss-Newton otherwise), instead of the backcasting backward
pass. Direct port of ``R/adam-gradient.R``; see also ``tests/testthat/
test_gradient.R`` on the R side.
"""

from __future__ import annotations

import numpy as np

from smooth import ADAM

# Seasonal series with trend — the case where backcasting historically drifted.
_Y = np.array(
    [
        10,
        12,
        15,
        13,
        16,
        18,
        20,
        19,
        22,
        25,
        28,
        30,
        11,
        13,
        16,
        14,
        17,
        19,
        21,
        20,
        23,
        26,
        29,
        31,
        12,
        14,
        17,
        15,
        18,
        20,
        22,
        21,
        24,
        27,
        30,
        32,
        13,
        15,
        18,
        16,
        19,
        21,
        23,
        22,
        25,
        28,
        31,
        33,
    ],
    dtype=float,
)


def test_gradient_runs_and_is_finite():
    m = ADAM(model="AAA", lags=[1, 12], initial="gradient").fit(_Y)
    fitted = np.asarray(m.fitted).ravel()
    assert np.all(np.isfinite(fitted))
    assert m.initial_type == "gradient"


def test_gradient_counts_initial_df_like_optimal():
    # Gradient solves the initials, so it spends the same degrees of freedom on
    # them as initial="optimal" (unlike the historical backcast zero-df count).
    mg = ADAM(model="AAA", lags=[1, 12], initial="gradient").fit(_Y)
    mo = ADAM(model="AAA", lags=[1, 12], initial="optimal").fit(_Y)
    assert mg.nparam == mo.nparam


def test_backcasting_counts_initial_df_like_optimal():
    # Initial states count towards df identically however they are obtained:
    # backcasting, gradient and optimal all report the same number of parameters
    # for the same model (structural df).
    mb = ADAM(model="AAA", lags=[1, 12], initial="backcasting").fit(_Y)
    mo = ADAM(model="AAA", lags=[1, 12], initial="optimal").fit(_Y)
    mg = ADAM(model="AAA", lags=[1, 12], initial="gradient").fit(_Y)
    assert mb.nparam == mo.nparam
    assert mg.nparam == mo.nparam


def test_gradient_additive_is_least_squares_optimum():
    # For an additive model the residuals are affine in the initial state, so the
    # one-shot solve lands at the global SSE minimum: no perturbation of the
    # initials at the fitted persistence can lower the in-sample SSE.
    m = ADAM(model="AAN", lags=[1], initial="gradient").fit(_Y)
    fitted = np.asarray(m.fitted).ravel()
    sse = float(np.sum((_Y - fitted) ** 2))
    # Compare against backcasting at the same spec — gradient should not be worse.
    mb = ADAM(model="AAN", lags=[1], initial="backcasting").fit(_Y)
    sse_b = float(np.sum((_Y - np.asarray(mb.fitted).ravel()) ** 2))
    assert sse <= sse_b + 1e-6


def test_gradient_multiplicative_runs():
    m = ADAM(model="MAM", lags=[1, 12], initial="gradient").fit(_Y)
    fitted = np.asarray(m.fitted).ravel()
    assert np.all(np.isfinite(fitted))
    assert np.all(fitted > 0)


def test_gradient_arima_falls_back_gracefully():
    # ARIMA is out of scope for the gradient solve; it must fall back to
    # backcasting rather than error.
    m = ADAM(
        model="NNN",
        orders={"ar": [1], "i": [1], "ma": [1]},
        initial="gradient",
    ).fit(_Y)
    assert np.all(np.isfinite(np.asarray(m.fitted).ravel()))


def test_gradient_nonseasonal():
    m = ADAM(model="ANN", lags=[1], initial="gradient").fit(_Y)
    assert np.all(np.isfinite(np.asarray(m.fitted).ravel()))


def test_gradient_msarima_solves_additive_arima_initials():
    # MSARIMA is an additive SSOE model, so initial="gradient" profiles the
    # ARIMA initials by least squares (matching optimal at fixed dynamics and
    # differing from backcasting), same as the R side.
    from smooth import MSARIMA

    def sse(m):
        return float(np.sum(np.asarray(m.residuals) ** 2))

    af = {"ar": [0.4], "ma": [0.3]}
    kw = dict(orders={"ar": [1], "i": [1], "ma": [1]}, lags=[1], arma=af, loss="MSE")
    mo = MSARIMA(initial="optimal", **kw).fit(_Y)
    mb = MSARIMA(initial="backcasting", **kw).fit(_Y)
    mg = MSARIMA(initial="gradient", **kw).fit(_Y)
    assert abs(sse(mg) - sse(mo)) < 1e-2 * max(1.0, sse(mo))
    assert abs(sse(mg) - sse(mb)) > 1e-5


def test_gradient_ces_solves_additive_ssoe_initials():
    # CES is an additive SSOE model, so initial="gradient" profiles the initials
    # by least squares (residuals are affine in the initial profile): it must run
    # (not fall back), differ from backcasting, and not lose on in-sample SSE.
    from smooth import CES

    def sse(m):
        return float(np.sum(np.asarray(m.residuals) ** 2))

    mg = CES(seasonality="none", initial="gradient", loss="MSE").fit(_Y)
    mb = CES(seasonality="none", initial="backcasting", loss="MSE").fit(_Y)
    assert np.isfinite(sse(mg))
    assert abs(sse(mg) - sse(mb)) > 1e-6  # genuinely solved, not a fallback
    assert sse(mg) <= sse(mb) * 1.001 + 1e-6  # no worse than backcasting


# --- Loss-aware solve -------------------------------------------------------

_RNG = np.random.default_rng(41)
_TREND = 100 + np.cumsum(_RNG.normal(0, 2, 72))
_SEAS = np.tile([5, -3, 2, -4, 6, -6, 3, -1, 2, -2, 4, -6], 6)
_YL = _TREND + _SEAS
_YLM = _TREND * np.tile(
    [1.05, 0.97, 1.02, 0.96, 1.06, 0.94, 1.03, 0.99, 1.02, 0.98, 1.04, 0.94], 6
)


def _loss_value(model):
    return float(model._adam_estimated["CF_value"])


def test_gradient_profiles_robust_one_step_losses():
    for loss in ["MAE", "HAM"]:
        mb = ADAM(model="AAA", lags=[12], initial="backcasting", loss=loss).fit(_YL)
        mg = ADAM(model="AAA", lags=[12], initial="gradient", loss=loss).fit(_YL)
        assert _loss_value(mg) < _loss_value(mb)


def test_gradient_profiles_multiplicative_likelihoods():
    for dist in [None, "dlnorm", "dinvgauss"]:
        kw = {} if dist is None else {"distribution": dist}
        mb = ADAM(model="MAM", lags=[12], initial="backcasting", **kw).fit(_YLM)
        mg = ADAM(model="MAM", lags=[12], initial="gradient", **kw).fit(_YLM)
        assert _loss_value(mg) < _loss_value(mb) + 1e-8


def test_gradient_solves_additive_multistep_losses():
    for loss in ["MSEh", "TMSE", "GTMSE", "MSCE", "GPL"]:
        mb = ADAM(model="AAA", lags=[12], initial="backcasting", loss=loss, h=6).fit(
            _YL
        )
        mg = ADAM(model="AAA", lags=[12], initial="gradient", loss=loss, h=6).fit(_YL)
        assert _loss_value(mg) < _loss_value(mb)


def test_gradient_loss_code_mapping_matches_r():
    # The mapping table must stay mirror-identical to adam_gradientLossCode in
    # R/adam-gradient.R -- this pins the full grid.
    from smooth.adam_general.core.utils.gradient import adam_gradient_loss_code

    grid = [
        (("MSE", "default", "A", None, 0, False), ("S", 0.0)),
        (("MAE", "default", "A", None, 0, False), ("A", 0.0)),
        (("HAM", "default", "A", None, 0, False), ("H", 0.0)),
        (("likelihood", "default", "A", None, 0, False), ("S", 0.0)),
        (("likelihood", "default", "M", None, 0, False), ("g", 0.0)),
        (("likelihood", "dlaplace", "A", None, 0, False), ("A", 0.0)),
        (("likelihood", "ds", "M", None, 0, False), ("H", 0.0)),
        (("likelihood", "dgnorm", "A", 1.5, 0, False), ("G", 1.5)),
        (("likelihood", "dgnorm", "A", None, 0, False), ("S", 0.0)),
        (("likelihood", "dlnorm", "M", None, 0, False), ("l", 0.0)),
        (("likelihood", "dlnorm", "A", None, 0, False), ("S", 0.0)),
        (("likelihood", "dgamma", "M", None, 0, False), ("g", 0.0)),
        (("likelihood", "dinvgauss", "M", None, 0, False), ("i", 0.0)),
        (("LASSO", "default", "A", None, 0, False), ("S", 0.0)),
        (("MSEh", "default", "A", None, 6, True), ("h", 6.0)),
        (("TMSE", "default", "A", None, 6, True), ("T", 6.0)),
        (("GTMSE", "default", "A", None, 6, True), ("t", 6.0)),
        (("MSCE", "default", "A", None, 6, True), ("C", 6.0)),
        (("GPL", "default", "A", None, 6, True), ("P", 6.0)),
        (("MAEh", "default", "A", None, 6, True), ("S", 0.0)),
    ]
    for args, expected in grid:
        assert adam_gradient_loss_code(*args) == expected
    assert adam_gradient_loss_code("custom", "default", "A", None, 0, False) is None
    assert (
        adam_gradient_loss_code(lambda a, f, b: 0.0, "default", "A", 0, 0, False)
        is None
    )


def test_gradient_custom_loss_falls_back_to_backcasting():
    def loss_fn(actual, fitted, B):  # noqa: N803
        return float(np.sum(np.abs(actual - fitted) ** 1.5))

    mb = ADAM(model="AAA", lags=[12], initial="backcasting", loss=loss_fn).fit(_YL)
    mg = ADAM(model="AAA", lags=[12], initial="gradient", loss=loss_fn).fit(_YL)
    assert abs(_loss_value(mg) - _loss_value(mb)) < 1e-8


# --- Occurrence models (om) -------------------------------------------------

_OT = np.array(
    [
        (1 if v > 0.5 else 0)
        for v in (0.5 + 0.4 * np.sin(np.arange(120) / 7) + 0.003 * np.arange(120))
    ],
    dtype=float,
)


def test_gradient_om_reaches_optimal_quality():
    from smooth import OM

    for occ in ["odds-ratio", "inverse-odds-ratio", "direct"]:
        mb = OM(model="MNN", occurrence=occ, initial="backcasting").fit(_OT)
        mo = OM(model="MNN", occurrence=occ, initial="optimal").fit(_OT)
        mg = OM(model="MNN", occurrence=occ, initial="gradient").fit(_OT)
        assert np.isfinite(float(mg.loglik))
        # Gradient profiles the same likelihood over the initials: at least
        # backcasting, and within 1% of the fully optimal fit.
        assert float(mg.loglik) >= float(mb.loglik) - 1e-4
        assert abs(float(mg.loglik) - float(mo.loglik)) < 0.01 * abs(float(mo.loglik))


def test_gradient_om_loss_code_mapping_matches_r():
    from smooth.adam_general.core.utils.gradient import adam_gradient_om_loss_code

    assert adam_gradient_om_loss_code("likelihood") == ("B", 0.0)
    assert adam_gradient_om_loss_code("MSE") == ("S", 0.0)
    assert adam_gradient_om_loss_code("MAE") == ("A", 0.0)
    assert adam_gradient_om_loss_code("HAM") == ("H", 0.0)
    assert adam_gradient_om_loss_code("LASSO") == ("S", 0.0)
    assert adam_gradient_om_loss_code("custom") is None
    assert adam_gradient_om_loss_code(lambda a, f, b: 0.0) is None


def test_gradient_om_fixed_persistence_loglik_is_finite():
    # Fixed persistence + gradient (or backcasting) leaves B empty: the
    # estimator must still evaluate the cost once (mirroring R's length(B)==0
    # guard) instead of returning nlopt's uninitialised +inf. Regression for
    # the -inf logLik bug.
    from smooth import OM

    for ini in ["backcasting", "gradient"]:
        m = OM(model="MNN", occurrence="odds-ratio", initial=ini, persistence=0.1).fit(
            _OT
        )
        assert len(m._adam_estimated["B"]) == 0
        ll = float(m.loglik)
        assert np.isfinite(ll)
        # logLik must equal the Bernoulli likelihood of the fitted probabilities
        f = np.asarray(m.fitted).ravel()
        on, off = f[_OT == 1], f[_OT == 0]
        expected = float(np.sum(np.log(on)) + np.sum(np.log(1 - off)))
        assert abs(ll - expected) < 1e-6


def test_om_loglik_is_bernoulli_not_cost_for_non_likelihood_loss():
    # For a fit-only loss (MSE), logLik must be the Bernoulli likelihood of the
    # fitted probabilities, NOT -CF (the MSE). lossValue keeps the MSE. No
    # flooring of the probabilities.
    from smooth import OM

    m = OM(model="MNN", occurrence="odds-ratio", initial="optimal", loss="MSE").fit(_OT)
    f = np.asarray(m.fitted).ravel()
    with np.errstate(invalid="ignore", divide="ignore"):
        bernoulli = float(np.sum(_OT * np.log(f) + (1 - _OT) * np.log(1 - f)))
    assert abs(float(m.loglik) - bernoulli) < 1e-9
    # lossValue is the MSE, a small number, not the Bernoulli magnitude
    assert m._adam_estimated["CF_value"] < 1.0


def test_om_lossvalue_is_the_fitting_loss_even_when_infeasible():
    # lossValue must be the actual fitting loss on the fitted probabilities,
    # never the optimiser's infeasibility penalty (1e300) — even when the loss
    # optimum leaves [0, 1]. Constructed data where odds-ratio MAE goes
    # infeasible (p < 0), so the guard would otherwise fire.
    from smooth import OM

    ot = np.tile([0.0, 0.0, 1.0, 0.0, 1.0, 0.0], 20)
    m = OM(
        model="MNN",
        occurrence="odds-ratio",
        initial="gradient",
        persistence=0.1,
        loss="MAE",
    ).fit(ot)
    f = np.asarray(m.fitted).ravel()
    assert abs(float(m.loss_value) - float(np.mean(np.abs(ot - f)))) < 1e-9
    assert float(m.loss_value) < 1.0  # the MAE, not 1e300
    # If the fit is infeasible (some p outside [0, 1]) the Bernoulli logLik is
    # NaN/-inf — surfaced, not clipped; if feasible it is finite. Either way it
    # is decoupled from lossValue.
    if f.min() < 0 or f.max() > 1:
        assert not np.isfinite(float(m.loglik))
